### DISCUSSION TITLE
Issue with undefined references when compiling LIS

### DISCUSSION THREAD

**Discussion by mcthreems** [NASA-LIS/LISF#1657] [Q&A] [answered]
Hello, I am attempting to compile LIS and have gotten to what seems like the last step in the compilation process. The compilation fails here due to several undefined references in some fortran files (the terminal output is attached as "comp_out.txt"). Many of these references include "mpi" in their name, so it might be an issue with the mpi module. Do I need to somehow point the compilation to the mpi module directly or is the issue elsewhere?
[comp_out.txt](https://github.com/user-attachments/files/18158306/comp_out.txt)

**Comment by jvgeiger** [comment]
Hello,

I see an inconsistency in your comp_out.txt file regarding MPI.

First, you must set the compiler environment variables like this:

```
export LIS_FC=mpif90
export LIS_CC=mpicc
```


Second, please check your ESMF library.  You are using /lfs/matthewmckinney.ui/LISF/esmf-8.7.0/lib/libO/Linux.gfortran.64.mpiuni.default.
The “mpiuni” says that ESMF was installed without MPI support.  Please check your /lfs/matthewmckinney.ui/LISF/esmf-8.7.0/lib/libO directory.  Do you have a subdirectory named something like Linux.gfortran.64.mpich3.default?  If no, then you must reinstall ESMF with MPI support enabled.  If yes, then please fix the `LIS_MODESMF` and `LIS_LIBESMF` environment variables to use it.


Then please go into your LISF/lis directory.  Run

```
make -C make realclean
./configure
./compile
```

When you run the configure utility, make sure to enable MPI support in LIS:

```
Parallelism (0-serial, 1-dmpar, default=1): 1
```


You also have a problem with HDF-EOS2.  Are you reading data in the HDF4 and/or HDF-EOS2 format?



Thanks,
Jim

**Reply by mcthreems** [reply]
Thanks for the suggestions. Regarding HDF4/HDF-EOS2, I don't know what specific files I may need to use with the model so wanted to include more options, but if these are causing an issue I can compile without them.

**Reply by jvgeiger** [reply]
First, let's double check that your HDFEOS2 library is in /lfs/matthewmckinney.ui/miniforge3/lib.  You are looking for both libhdfeos.a and libGctp.a.  If not, then you have to fix your `LIS_HDFEOS` environment variable.

**Reply by mcthreems** [reply]
Yes, both of those files are in the /lfs/matthewmckinney.ui/miniforge3/lib directory.

**Reply by jvgeiger** [reply]
Let's first fix your MPI issue.

**Reply by mcthreems** [reply]
I made the changes you recommended for the mpi issue and when doing the compilation I now get different undefined references, all of which start with "gd" as in "read_ANSASNWDsnow.F90:(.text+0x63f): undefined reference to `gdopen_'".

**Reply by jvgeiger** [reply]
That is HDFEOS.

Let's make sure that you have no other problems.  Please disable HDF4 and HDFEOS support.

```
make -C make realclean
./configure
./compile
```

When you run the configure utility, respond with:

```
Use HDF4? (1-yes, 0-no, default=1): 0
Use HDFEOS? (1-yes, 0-no, default=1): 0
```

**Reply by mcthreems** [reply]
Disabling HDF4 and HDFEOS did allow the compilation to complete, so it seems like those are the only remaining issue. I assume I can use the current compiled version as long as I am not trying to use those particular file types?

**Reply by jvgeiger** [reply]
Yes, you are correct.  That LIS executable is ready for use (without HDF4 or HDFEOS2).

If you want to fix the HDFEOS error, then I suggest that you first enable just HDF4.

```
make -C make realclean
./configure
./compile
```

When you run the configure utility, respond with:

```
Use HDF4? (1-yes, 0-no, default=1): 1
Use HDFEOS? (1-yes, 0-no, default=1): 0
```

**Reply by mcthreems** [reply]
Enabling just hdf4 does result in a small number of undefined references, which start with "sf" such as in "read_PMW_snow.F90:(.text+0xf06): undefined reference to `sfstart_'". So it seems like both HDF4 and HDFEOS have issues.

**Reply by mcthreems** [reply]
If I want to try a test case for a compilation without HDF4 and HDFEOS, which one would be best suited?

**Reply by mcthreems** [reply]
Hello Jim,

I have been able to successfully compile with HDF4 disabled, so now just need to resolve the issues with that. Currently the issues appear to be the same as in the previous comment, that being a handful of undefined references such as 'sfstart_' in the file read_PMW_snow.F90.

**Reply by jvgeiger** [reply]
Hello,

What does `echo $LIS_HDF4` report?  And where is your HDF4 library installed?

**Reply by mcthreems** [reply]
The output of "echo $LIS_HDF4" is "/lfs/software/spack-jb/opt/spack/linux-rocky8-broadwell/gcc-12.1.0/hdf-4.2.15-qlpnnqcfr3e332jlvvdripw5h3g6olc3", which should match the installation of HDF4 on our cluster.

**Reply by jvgeiger** [reply]
Let's make sure that everything was installed correctly.  What does `ls /lfs/software/spack-jb/opt/spack/linux-rocky8-broadwell/gcc-12.1.0/hdf-4.2.15-qlpnnqcfr3e332jlvvdripw5h3g6olc3` report? And what does ls /lfs/software/spack-jb/opt/spack/linux-rocky8-broadwell/gcc-12.1.0/hdf-4.2.15-qlpnnqcfr3e332jlvvdripw5h3g6olc3/lib` report?

**Reply by mcthreems** [reply]
The first command outputs "bin  include  lib  share", and the second one outputs "libdf.a  libhdf4.settings  libmfhdf.a".

**Reply by jvgeiger** [reply]
That looks correct.

Please check the make/configure.lis file.  Make sure that the LIB_HDF5 and INC_HDF5 variables correctly points to your HDF5 library.  If not, then you should rerun the configure and compile scripts:

```
make -C make realclean
./configure
./compile -j 16
```

**Reply by mcthreems** [reply]
Those variables do point to the correct paths, so should not be causing the issues. I currently speculate that our issue is conflicting installs of the same modules (netcdf, hdf4, etc) between my own account on our cluster and cluster-wide installs, so I am currently trying to fully switch over to using only cluster-wide modules. Once we get those set up and I attempt a compilation with them I will let you know if there are still issues.

**Reply by mcthreems** [reply]
We have switched over to cluster-wide installs of the necessary modules, but are having some issues with the LD flags. Specifically the flags "-lnetcdf", "-ljpeg", and "-lhdf5_hl" seem to refer to different versions of modules than the ones we are using, but removing them results in several undefined reference errors.

**Comment by jvgeiger** [comment]
Hello.  All I can think of right now is to check your environment settings.

Examples:

1. Make sure that your modules are setting the correct paths.

2. Make sure that your LD_LIBRARY_PATH is set correctly.  Do you have multiple different entries for the same library?  I write a custom modulefile for managing my development/running environment for LIS.  In it, I load all needed modules and set all needed environment variables.  I do *not* set any of these in my .profile file.

3. Make sure that all your libraries were installed consistently.  Make sure that your libraries were compiled to use the same dependencies.  I have seen issues when using a mix of libraries compiled manually with libraries installed via spack.  spack will install dependecies, so lib1 installed manually and lib2 installed with spack end up referring to two different versions of a lowel level dependency.

**Reply by mcthreems** [reply]
Thank you for the suggestions. We have been working on this on our end as well and have managed to get the compilation to finish, but need to test whether or not it actually works. Is there a simple test we can perform to check if our LIS executable is working?

**Reply by mcthreems** [reply, marked as answer]
To finish this thread, the original issue was resolved by both fixing the compiler environment variables and then tracking down which modules included the references that the compiler was unable to find.

**Comment by jvgeiger** [comment]
Hello,

You can try our Public Testcase Walkthrough:

https://nasa-lis.github.io/LISF/public_testcase_walkthrough/public_testcase_walkthrough.html
